<?xml version="1.0"?>
<robot name="whac-a-mole">

    <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
        <robotNamespace>/</robotNamespace>
        <allow_free_joints>true</allow_free_joints>
    </plugin>
    </gazebo>

    <link name="world" />

    <joint name="world_joint"  type="fixed">
        <origin xyz="5.5 4 0" rpy="0 0 0" />
        <parent link="world"/>
        <child link="base"/>
    </joint>

    <link name="base">
        <visual>
            <geometry>
                <mesh filename="package://whac-a-mole_description/meshes/Base.stl"/>
            </geometry>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <!-- <collision>
            <geometry>
                <mesh filename="package://whac-a-mole_description/meshes/Base.stl"/>
            </geometry>
        </collision> -->
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.001"/>
        </inertial>
    </link>

    <!-- Linear Actuator 1 -->
    <link name="linear_actuator_1">
        <visual>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.001"/>
        </inertial>
    </link>

    <link name="slider_link_1">
        <visual>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 0.8 1"/>
            </material>
        </visual>
        <!-- <collision>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
        </collision> -->
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.0001"/>
        </inertial>
    </link>

    <joint name="slider_joint_1" type="prismatic">
        <parent link="world"/>
        <child link="slider_link_1"/>
        <origin xyz="4 -4 4.75" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="100.0" lower="-2.25" upper="0.4" velocity="1.0"/>
        <dynamics damping="0.7" friction="0.0"/>
    </joint>
    
    <joint name="base_to_slider_link_1" type="fixed">
        <parent link="slider_link_1"/>
        <child link="linear_actuator_1"/>
        <origin xyz="0 0 -1.75" rpy="0 0 0"/>
    </joint>

    <!-- Linear Actuator 2 -->
    <link name="linear_actuator_2">
        <visual>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.001"/>
        </inertial>
    </link>

    <link name="slider_link_2">
        <visual>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 0.8 1"/>
            </material>
        </visual>
        <!-- <collision>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
        </collision> -->
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.0001"/>
        </inertial>
    </link>

    <joint name="slider_joint_2" type="prismatic">
        <parent link="world"/>
        <child link="slider_link_2"/>
        <origin xyz="2 -4 4.75" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="100.0" lower="-2.25" upper="0.4" velocity="1.0"/>
        <dynamics damping="0.7" friction="0.0"/>
    </joint>
    
    <joint name="base_to_slider_link_2" type="fixed">
        <parent link="slider_link_2"/>
        <child link="linear_actuator_2"/>
        <origin xyz="0 0 -1.75" rpy="0 0 0"/>
    </joint>

    <!-- Linear Actuator 3 -->
    <link name="linear_actuator_3">
        <visual>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.001"/>
        </inertial>
    </link>

    <link name="slider_link_3">
        <visual>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 0.8 1"/>
            </material>
        </visual>
        <!-- <collision>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
        </collision> -->
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.0001"/>
        </inertial>
    </link>

    <joint name="slider_joint_3" type="prismatic">
        <parent link="world"/>
        <child link="slider_link_3"/>
        <origin xyz="-1 -4 4.75" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="100.0" lower="-2.25" upper="0.4" velocity="1.0"/>
        <dynamics damping="0.7" friction="0.0"/>
    </joint>
    
    <joint name="base_to_slider_link_3" type="fixed">
        <parent link="slider_link_3"/>
        <child link="linear_actuator_3"/>
        <origin xyz="0 0 -1.75" rpy="0 0 0"/>
    </joint>

    <!-- Linear Actuator 4 -->
    <link name="linear_actuator_4">
        <visual>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
            <material name="gray">
                <color rgba="0.5 0.5 0.5 1"/>
            </material>
        </visual>
        <collision>
            <geometry>
                <box size="1 1 0.5"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.001"/>
        </inertial>
    </link>

    <link name="slider_link_4">
        <visual>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 0.8 1"/>
            </material>
        </visual>
        <!-- <collision>
            <geometry>
                <cylinder radius="0.2" length="3.00"/>
            </geometry>
        </collision> -->
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" 
                     iyy="0.001" iyz="0.0" 
                     izz="0.0001"/>
        </inertial>
    </link>

    <joint name="slider_joint_4" type="prismatic">
        <parent link="world"/>
        <child link="slider_link_4"/>
        <origin xyz="-3 -4 4.75" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit effort="100.0" lower="-2.25" upper="0.4" velocity="1.0"/>
        <dynamics damping="0.7" friction="0.0"/>
    </joint>
    
    <joint name="base_to_slider_link_4" type="fixed">
        <parent link="slider_link_4"/>
        <child link="linear_actuator_4"/>
        <origin xyz="0 0 -1.75" rpy="0 0 0"/>
    </joint>

        <!-- Transmission for slider_joint_1 -->
    <transmission name="slider_joint_1_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="slider_joint_1">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="slider_joint_1_motor">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- Transmission for slider_joint_2 -->
    <transmission name="slider_joint_2_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="slider_joint_2">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="slider_joint_2_motor">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- Transmission for slider_joint_3 -->
    <transmission name="slider_joint_3_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="slider_joint_3">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="slider_joint_3_motor">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- Transmission for slider_joint_4 -->
    <transmission name="slider_joint_4_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="slider_joint_4">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="slider_joint_4_motor">
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- Button 1 -->
    <!-- Base Link -->
    <link name="button_base_1">
        <visual>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
            <material name="gray"/>
        </visual>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0"
                     iyy="0.0001" iyz="0"
                     izz="0.0001"/>
        </inertial>
    </link>
    
    <!-- Button Link -->
    <link name="button_top_1">
        <visual>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.00005" ixy="0" ixz="0"
                     iyy="0.00005" iyz="0"
                     izz="0.00005"/>
        </inertial>
    </link>
    
    <!-- Prismatic Joint -->
    <joint name="button_joint_1" type="prismatic">
        <parent link="button_base_1"/>
        <child link="button_top_1"/>
        <origin xyz="0 0 1.65" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1" upper="0" effort="10" velocity="0.1"/>
        <dynamics damping="0.1" friction="0.0"/>
    </joint>
    
    <!-- Fixed Joint to Attach Button Base to World -->
    <joint name="button_1_base_to_world" type="fixed">
        <parent link="world"/>
        <child link="button_base_1"/>
        <origin xyz="4 -4 0.5" rpy="0 0 0"/>
    </joint>
    
    <!-- Gazebo Extension for Spring Effect -->
    <gazebo reference="button_joint_1">
        <spring_stiffness>100.0</spring_stiffness>
        <spring_reference>0.0</spring_reference>
        <damping>0.5</damping>
    </gazebo>

    <!-- Button 2 -->
    <!-- Base Link -->
    <link name="button_base_2">
        <visual>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
            <material name="gray"/>
        </visual>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0"
                     iyy="0.0001" iyz="0"
                     izz="0.0001"/>
        </inertial>
    </link>
    
    <!-- Button Link -->
    <link name="button_top_2">
        <visual>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.00005" ixy="0" ixz="0"
                     iyy="0.00005" iyz="0"
                     izz="0.00005"/>
        </inertial>
    </link>
    
    <!-- Prismatic Joint -->
    <joint name="button_joint_2" type="prismatic">
        <parent link="button_base_2"/>
        <child link="button_top_2"/>
        <origin xyz="0 0 1.65" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1" upper="0" effort="10" velocity="0.1"/>
        <dynamics damping="0.1" friction="0.0"/>
    </joint>
    
    <!-- Fixed Joint to Attach Button Base to World -->
    <joint name="button_2_base_to_world" type="fixed">
        <parent link="world"/>
        <child link="button_base_2"/>
        <origin xyz="2 -4 0.5" rpy="0 0 0"/>
    </joint>
    
    <!-- Gazebo Extension for Spring Effect -->
    <gazebo reference="button_joint_2">
        <spring_stiffness>100.0</spring_stiffness>
        <spring_reference>0.0</spring_reference>
        <damping>0.5</damping>
    </gazebo>

    <!-- Button 3 -->
    <!-- Base Link -->
    <link name="button_base_3">
        <visual>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
            <material name="gray"/>
        </visual>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0"
                     iyy="0.0001" iyz="0"
                     izz="0.0001"/>
        </inertial>
    </link>
    
    <!-- Button Link -->
    <link name="button_top_3">
        <visual>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
            <material name="blue">
                <color rgba="0.0 0.0 0.8 1"/>
                <ambient>0 0 0 1</ambient>
                <diffuse>0 0 0 1</diffuse>
                <specular>0 0 0 0</specular>
                <emissive>0 0 1 1</emissive>
            </material>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.00005" ixy="0" ixz="0"
                     iyy="0.00005" iyz="0"
                     izz="0.00005"/>
        </inertial>
    </link>
    
    <!-- Prismatic Joint -->
    <joint name="button_joint_3" type="prismatic">
        <parent link="button_base_3"/>
        <child link="button_top_3"/>
        <origin xyz="0 0 1.65" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1" upper="0" effort="10" velocity="0.1"/>
        <dynamics damping="0.1" friction="0.0"/>
    </joint>
    
    <!-- Fixed Joint to Attach Button Base to World -->
    <joint name="button_3_base_to_world" type="fixed">
        <parent link="world"/>
        <child link="button_base_3"/>
        <origin xyz="-1 -4 0.5" rpy="0 0 0"/>
    </joint>
    
    <!-- Gazebo Extension for Spring Effect -->
    <gazebo reference="button_joint_3">
        <spring_stiffness>100.0</spring_stiffness>
        <spring_reference>0.0</spring_reference>
        <damping>0.5</damping>
    </gazebo>

    <!-- Button 4 -->
    <!-- Base Link -->
    <link name="button_base_4">
        <visual>
            <geometry>
                <box size="1 1 1"/>
            </geometry>
            <material name="gray"/>
        </visual>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.0001" ixy="0" ixz="0"
                     iyy="0.0001" iyz="0"
                     izz="0.0001"/>
        </inertial>
    </link>
    
    <!-- Button Link -->
    <link name="button_top_4">
        <visual>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.45" length="0.3"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.1"/>
            <inertia ixx="0.00005" ixy="0" ixz="0"
                     iyy="0.00005" iyz="0"
                     izz="0.00005"/>
        </inertial>
    </link>
    
    <!-- Prismatic Joint -->
    <joint name="button_joint_4" type="prismatic">
        <parent link="button_base_4"/>
        <child link="button_top_4"/>
        <origin xyz="0 0 1.65" rpy="0 0 0"/>
        <axis xyz="0 0 1"/>
        <limit lower="-1" upper="0" effort="10" velocity="0.1"/>
        <dynamics damping="0.1" friction="0.0"/>
    </joint>
    
    <!-- Fixed Joint to Attach Button Base to World -->
    <joint name="button_4_base_to_world" type="fixed">
        <parent link="world"/>
        <child link="button_base_4"/>
        <origin xyz="-3 -4 0.5" rpy="0 0 0"/>
    </joint>
    
    <!-- Gazebo Extension for Spring Effect -->
    <gazebo reference="button_joint_4">
        <spring_stiffness>100.0</spring_stiffness>
        <spring_reference>0.0</spring_reference>
        <damping>0.5</damping>
    </gazebo>
</robot>